joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -85,31 +90,154 @@ JointTrajectoryController::update()
+          //break;
+        }
+      }
 
-    size_t joint_num = registered_joint_cmd_handles_.size();
-    for (size_t index = 0; index < joint_num; ++index) {
-      registered_joint_cmd_handles_[index]->set_cmd(traj_point_ptr->positions[index]);
+      if (rt_active_goal_) {
+        // send feedback
+        auto feedback = std::make_shared<FollowJTrajAction::Feedback>();
+        feedback->header.stamp = lifecycle_node_->now();
+        feedback->joint_names = joint_names_;
+
+        feedback->actual = state_current;
+        feedback->desired = state_desired;
+        feedback->error = state_error;
+        rt_active_goal_->setFeedback(feedback);
+
+        // check abort
+        if (abort) {
+          // RCLCPP_WARN(lifecycle_node_->get_logger(), "Aborted due to state tolerance violation");
+          // auto result = std::make_shared<FollowJTrajAction::Result>();
+          // result->set__error_code(FollowJTrajAction::Result::PATH_TOLERANCE_VIOLATED);
+          // rt_active_goal_->setAborted(result);
+          // rt_active_goal_.reset();
+        }
+
+        // check goal tolerance
+        if (!before_last_point) {
+          if (!outside_goal_tolerance) {
+            auto res = std::make_shared<FollowJTrajAction::Result>();
+            res->set__error_code(FollowJTrajAction::Result::SUCCESSFUL);
+            rt_active_goal_->setSucceeded(res);
+            rt_active_goal_.reset();
+
+            RCLCPP_INFO(lifecycle_node_->get_logger(), "Goal reached, success!");
+          } else if (default_tolerances_.goal_time_tolerance != 0.0) {
+            // if we exceed goal_time_toleralance set it to aborted
+            rclcpp::Time traj_start = (*traj_point_active_ptr_)->get_trajectory_start_time();
+            rclcpp::Time traj_end = traj_start + start_segment_itr->time_from_start;
+
+            double difference = lifecycle_node_->now().seconds() - traj_end.seconds();
+            if (difference > default_tolerances_.goal_time_tolerance) {
+              // auto result = std::make_shared<FollowJTrajAction::Result>();
+              // result->set__error_code(FollowJTrajAction::Result::GOAL_TOLERANCE_VIOLATED);
+              // rt_active_goal_->setAborted(result);
+              // rt_active_goal_.reset();
+              /*RCLCPP_WARN(
+                lifecycle_node_->get_logger(),
+                "Aborted due goal_time_tolerance exceeding by %f seconds (given: %f)",
+                difference, default_tolerances_.goal_time_tolerance);*/
+            }
+          }
+        }
+      }
     }
 
-    prev_traj_point_ptr_ = traj_point_ptr;
     set_op_mode(hardware_interface::OperationMode::ACTIVE);
   }
+  else
+  {
+    state_desired = traj_msg_home_ptr_->points[0];
 
-  publish_state();
+    for (auto index = 0ul; index < joint_num; ++index) {
+        compute_error_for_joint(state_error, index, state_current, state_desired);
+    }
+  }
+  
+  double period = lifecycle_node_->get_parameter("period").as_double();
```

Comment:
This hack is supplied upstream from gazebo_ros_controls; ideally we should use something else.